docs/core/common/assets/js/api/core/ui-states.md

```diff
@@ -0,0 +1,178 @@
+## API - `$e.uiStates`
+*  **Description**: `$e.uiStates` API is a UI state manager that allows you to create custom UI states for your components.
+*  **Location**: *core/common/assets/js/api/core/ui-states.js*
+*  **Methods**:
+
+| Method                     | Params                                                   | Returns   | Description
+|----------------------------|----------------------------------------------------------|-----------|---------------------------------------------------------------------
+| `$e.uiStates.register()`   |`{UiStateBase}` *instance*   								| `{void}`  | Register a new UI state.
+| `$e.uiStates.getAll()`     |                                                          | `{Object}`| Get all existing UI states with their options.
+| `$e.uiStates.get()`        | `{String}` *state*                                       | `{Object}`| Get the state value, or return all of them if no `state` is set.
+| `$e.uiStates.set()`        | `{String}` *state*, `{String}` *value*                   | `{void}`  | Set the current state value and trigger its callbacks & events.
+| `$e.uiStates.remove()`     | `{String}` *state*                                       | `{void}`  | Remove a state.
+| `$e.uiStates.getCurrent()` | `{String}` *state*                                       | `{string}`| Get the current state value.
+* **Notes**:
+	- Each UI state has its own options & callback for each option.
+	- When a UI state is being changed, it:
+		- Adds a CSS class to the scope elements ( `e-ui-state--${ stateID }__${ value }` ) - Slashes are being replaced by hyphens ( e.g. `document/direction-mode` will become `document-direction-mode` ).
+		- Dispatches a custom event to the scope elements ( `e-ui-state:${ stateID }`, e.g. `e-ui-state:document/direction-mode` ) with `oldValue` and `newValue` under `e.detail`.
+	- Setting an invalid option to a state will throw an error.
+
+## Guidelines, Conventions & Files Structure:
+* Each UI state is owned by a [component](./components.md#guidelines-conventions--files-structure).
+* Each [component](./components.md#guidelines-conventions--files-structure) can override the `defaultUiStates()` method to return a set of UI states of its own.
+* The UI states are imported via built-in method called `importUiStates()`.
+	* All the UI states should be exported in one index file:
+	    ```javascript
+		// ui-states/index.js
+		export { DirectionMode } from './direction-mode';
+		export { ColorPickingMode } from './color-picking-mode';
+		```
+	  You can have as many indexes as you wish in every hierarchy under `component/ui-states/some/long/path` in order to organize your code, but the requirement is to have a single index file
```

Comment:
You can have as many indexes as you wish in every hierarchy under `component/ui-states/some/long/path` to organize